vault/seal_autoseal.go

```diff
@@ -499,3 +506,52 @@ func (d *autoSeal) migrateRecoveryConfig(ctx context.Context) error {
 
 	return nil
 }
+
+func (d *autoSeal) StartHealthCheck() {
+	d.healthCheck = time.NewTicker(sealHeathTestInterval)
+	d.healthCheckStop = make(chan struct{})
+	go func() {
+		lastTestOk := true
+		lastSeenOk := time.Now()
+		for {
+			select {
+			case <-d.healthCheckStop:
+				d.healthCheck.Stop()
+				close(d.healthCheckStop)
+				d.healthCheck = nil
+				d.healthCheckStop = nil
+				return
+			case t := <-d.healthCheck.C:
+				testVal := fmt.Sprintf("Heartbeat %d", mathrand.Intn(1000))
+				ciphertext, err := d.Wrapper.Encrypt(d.core.activeContext, []byte(testVal), nil)
+				if err != nil {
+					lastTestOk = false
+					d.logger.Warn("failed to encrypt seal health test value, seal backend may be unreachable", "error", err)
+				} else {
+					plaintext, err := d.Wrapper.Decrypt(d.core.activeContext, ciphertext, nil)
+					if err != nil {
+						lastTestOk = false
+						d.logger.Warn("failed to decrypt seal health test value, seal backend may be unreachable", "error", err)
+					}
+					if !bytes.Equal([]byte(testVal), plaintext) {
+						lastTestOk = false
+						d.logger.Warn("seal health test value failed to decrypt to expected value")
+					} else {
+						d.logger.Debug("seal health test passed")
+						if !lastTestOk {
+							d.logger.Info("seal backend is now healthy again", "downtime", t.Sub(lastSeenOk).String())
+						}
+						lastTestOk = true
+						lastSeenOk = t
+					}
+				}
+			}
+		}
+	}()
+}
+
+func (d *autoSeal) StopHealthCheck() {
+	if d.healthCheckStop != nil {
+		d.healthCheckStop <- struct{}{}
```

Comment:
Wouldn't it be safer to close the channel?  That way if we're blocked waiting for a response from the upstream seal service, we won't block sealing Vault.

Reply:
Good idea.